mobile/src/main/java/org/openhab/habdroid/core/connection/ConnectionFactory.java

```diff
@@ -0,0 +1,275 @@
+                    // a new callback.
+                    if (msg.obj == mLocalConnection || msg.obj == mRemoteConnection) {
+                        updateAvailableConnection((Connection) msg.obj, null);
+                    }
+                }
+                return true;
+        }
+        return false;
+    }
+
+    private void updateAvailableConnection(Connection c, ConnectionException failureReason) {
+        if (failureReason != null) {
+            mConnectionFailureReason = failureReason;
+            mAvailableConnection = null;
+        } else if (c == mAvailableConnection) {
+            return;
+        } else {
+            mConnectionFailureReason = null;
+            mAvailableConnection = c;
+        }
+        LocalBroadcastManager lbm = LocalBroadcastManager.getInstance(ctx);
+        lbm.sendBroadcast(new Intent(ACTION_NETWORK_CHANGED));
+    }
+
+    // called in update thread
+    private static Connection determineAvailableConnection(Context context,
+            Connection local, Connection remote) throws ConnectionException {
+        ConnectivityManager connectivityManager =
+                (ConnectivityManager) context.getSystemService(Context.CONNECTIVITY_SERVICE);
+        NetworkInfo info = connectivityManager.getActiveNetworkInfo();
+
+        if (info == null || !info.isConnected()) {
+            Log.e(TAG, "Network is not available");
+            throw new NetworkNotAvailableException();
+        }
+
+        // If we are on a mobile network go directly to remote URL from settings
+        if (info.getType() == ConnectivityManager.TYPE_MOBILE) {
+            if (remote == null) {
+                throw new NoUrlInformationException();
+            }
+            return remote;
+        }
+
+        // Else if we are on Wifi, Ethernet, WIMAX or VPN network
+        if (localConnectionTypes.contains(info.getType())) {
+            // If local URL is configured and rechable
+            if (local != null && local.checkReachabilityInBackground()) {
+                Log.d(TAG, "Connecting to local URL");
+
+                return local;
+            }
+            // If local URL is not reachable or not configured, try with remote URL
+            if (remote != null) {
+                Log.d(TAG, "Connecting to remote URL");
+                return remote;
+            } else {
+                throw new NoUrlInformationException();
+            }
+            // Else we treat other networks types as unsupported
+        } else {
+            Log.e(TAG, "Network type (" + info.getTypeName() + ") is unsupported");
+            throw new NetworkNotSupportedException(info);
+        }
+    }
+
+    @VisibleForTesting
+    public synchronized void updateConnections() {
+        if (settings.getBoolean(Constants.PREFERENCE_DEMOMODE, false)) {
+            mAvailableConnection = new DemoConnection(ctx, settings);
+            mLocalConnection = mRemoteConnection = mAvailableConnection;
```

Comment:
I wonder whether this (assigning mRemoteConnection) actually is correct. Does the demo server have an openhab-cloud instance running on it?

Reply:
Thinking about it it probably is, as the demo server is reachable from mobile network. My question above basically relates to the FIXME comment about REMOTE vs. CLOUD connection above. I have an idea for distinguishing both; will open a PR for that once I have it ready.